_posts/2019-06-05-uncertainty-estimates-tfprobability/uncertainty_tfprobability.Rmd

```diff
@@ -0,0 +1,782 @@
+
+```{r}
+
+model <- keras_model_sequential() %>%
+  layer_dense_variational(
+    units = 1,
+    make_posterior_fn = posterior_mean_field,
+    make_prior_fn = prior_trainable,
+    kl_weight = 1 / n
+  ) %>%
+  layer_distribution_lambda(function(x)
+    tfd_normal(loc = x, scale = 1))
+```
+
+
+You may have noticed one argument to `layer_dense_variational` we haven't discussed yet, `kl_weight`.
+This is used to scale the contribution to the total loss of the KL divergence, and normally should equal one over the number of data points.
+
+Training the model is straightforward. As users, we only specify the _negative log likelihood_ part of the loss; the KL divergence part is taken care of transparently by the framework.
+
+```{r}
+negloglik <- function(y, model) - (model %>% tfd_log_prob(y))
+model %>% compile(optimizer = optimizer_adam(lr = learning_rate), loss = negloglik)
+model %>% fit(x, y, epochs = 1000)
+```
+
+Because of the stochasticity inherent in a variational-dense layer, each time we call this model, we obtain different results: different normal distributions, in this case.
+To obtain the uncertainty estimates we're looking for, we therefore call the model a bunch of times -- 100, say:
+
+```{r}
+yhats <- purrr::map(1:100, function(x) model(tf$constant(x_test)))
+```
+
+We can now plot those 100 predictions -- lines, in this case, as there are no nonlinearities:
+
+```{r}
+means <-
+  purrr::map(yhats, purrr::compose(as.matrix, tfd_mean)) %>% abind::abind()
+
+lines <- data.frame(cbind(x_test, means)) %>%
+  gather(key = run, value = value,-X1)
+
+mean <- apply(means, 1, mean)
+
+ggplot(data.frame(x = x, y = y, mean = as.numeric(mean)), aes(x, y)) +
+  geom_point() +
+  geom_line(aes(x = x_test, y = mean), color = "violet", size = 1.5) +
+  geom_line(
+    data = lines,
+    aes(x = X1, y = value, color = run),
+    alpha = 0.3,
+    size = 0.5
+  ) +
+  theme(legend.position = "none")
+```
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Epistemic uncertainty on simulated data, using linear activation in the variational-dense layer."}
+knitr::include_graphics("images/g_epistemic_linear_kl150.png")
+```
+
+What we see here are essentially _different models_, consistent with the assumptions built into the architecture. What we're not accounting for is the spread in the data. Can we do both? We can; but first let's comment on a few choices that were made and see how they affect the results. 
+
+To prevent this post from growing to infinite size, we've refrained from performing a systematic experiment; please take what follows not as generalizable statements, but as _pointers to things you will want to keep in mind_ in your own ventures. Especially, each (hyper-)parameter is not an island; they could interact in unforeseen ways.
```

Comment:
I wonder if it would be worth having a subheading for this section on questions/things to pay attention to.

Reply:
I'd rather not over-stress these, for pragmatic reasons...